pkg/policy/manager.go

```diff
@@ -0,0 +1,455 @@
+	if diags != nil && diags.HasErrors() {
+		return nil, fmt.Errorf("failed to load policy file: %#v", diags.Errs())
+	}
+	policies, diagsDecode := parser.DecodePolicies(policiesRaw, diags)
+	if diagsDecode != nil && diagsDecode.HasErrors() {
+		return nil, fmt.Errorf("failed to parse policy file: %#v", diagsDecode.Errs())
+	}
+
+	// Acquire connection from the connection pool
+	conn, err := m.pool.Acquire(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to acquire connection from the connection pool: %s", err.Error())
+	}
+	defer conn.Release()
+
+	// Traverse all policies recursively
+	policyMap := make(map[string]*config.Policy)
+	policyMap = traversePolicies(policies.Policies, "", policyMap)
+
+	// No sub path provided. Execute everything.
+	executor := NewExecutor(conn)
+	if p.SubPath == "" {
+		return executor.ExecutePolicies(ctx, execReq, policyMap)
+	}
+
+	// If we are here, the user only wants to execute a sub policy/view/query so we have to
+	// find the corresponding element.
+
+	// If the given path points directly to a sub policy
+	if _, ok := policyMap[p.SubPath]; ok {
+		// Collect all sub policies from this policy
+		subPolicyMap := make(map[string]*config.Policy)
+		for k, v := range policyMap {
+			// Add policy if prefix is the same.
+			// This will also add the actual policy since k == p.SubPath.
+			if strings.HasPrefix(k, p.SubPath) {
+				subPolicyMap[k] = v
+			}
+		}
+		return executor.ExecutePolicies(ctx, execReq, subPolicyMap)
+	}
+
+	// Must be a query so get the policy path and the last element
+	pathSplit := strings.Split(p.SubPath, pathDelimiter)
+	if len(pathSplit) <= 1 {
+		// Sub path is malformed
+		return nil, fmt.Errorf("malformed sub path: %s", p.SubPath)
+	}
+
+	// Get the policy path and last element
+	policyPath := pathSplit[len(pathSplit)-2]
+	elementName := pathSplit[len(pathSplit)-1]
+
+	// Get query parent policy
+	parentPolicy, ok := policyMap[policyPath]
+	if !ok {
+		return nil, fmt.Errorf("cannot find sub query parent policy %s in %s", policyPath, p.SubPath)
+	}
+
+	for _, query := range parentPolicy.Queries {
+		if query.Name == elementName {
+			// Execute query
+			res, err := executor.ExecuteQuery(ctx, query)
+			if err != nil {
+				return nil, err
+			}
+
+			// Collect results
+			execResults := &ExecutionResult{
+				Passed:  true,
+				Results: make(map[string]*QueryResult),
+			}
+			collectExecutionResults(execResults, policyPath, res)
+			if execReq.UpdateCallback != nil {
+				execReq.UpdateCallback(query.Name, execResults.Passed)
+			}
+			return execResults, nil
+		}
+	}
+
+	return nil, fmt.Errorf("cannot find sub query %s in %s", elementName, p.SubPath)
+}
+
+// traversePolicies is a recursive function that traverses p until all policies are resolved.
+// All traversed policies gets stored into m where the policy level is used as a key.
+func traversePolicies(p []*config.Policy, levelPath string, m map[string]*config.Policy) map[string]*config.Policy {
```

Comment:
I would add some debug logging here if we ever want to understand what got added to the policy map etc'.

Reply:
Good idea 👍